src/memo/memo.ml

```diff
@@ -671,19 +568,92 @@ module M = struct
 end
 
 module State = M.State
-module Running_state = M.Running_state
 module Dep_node = M.Dep_node
 module Deps = M.Deps
 
-module Stack_frame_with_state = struct
+module Stack_frame_with_state : sig
+  type phase =
+    | Restore_from_cache
+    | Compute
+
+  type t
+
+  val to_dyn : t -> Dyn.t
+
+  (* CR-someday amokhov: Shall we stop reusing DAG nodes? It used to be good
+     optimisation, but now that we do not create a DAG node for every [Dep_node]
+     it's probably useless. Dropping it will require some refactoring though.
+
+     aalekseyev: yeah, this reuse of DAG nodes is suspicious and I wish we got
+     rid of it. *)
+  (* Create a new stack frame related to restoring or computing a [dep_node]. By
+     providing a [dag_node], it is possible to reuse it instead of creating a
+     new one. We currently reuse the same DAG node for [Restore_from_cache] and
+     [Compute] phases of the same [dep_node]. *)
+  val create :
+       dag_node:Dag.node Lazy.t
+    -> phase
+    -> dep_node:_ Dep_node_without_state.t
+    -> t
+
+  val dep_node : t -> Dep_node_without_state.packed
+
+  val dag_node : t -> Dag.node Lazy.t
+
+  (* This function accumulates dependencies of frames from the [Compute] phase.
+     Calling it for a [Restore_from_cache] frame is a no-op. *)
+  val add_dep : t -> dep_node:_ Dep_node.t -> unit
+
+  val deps_rev : t -> Dep_node.packed list
+
+  val children_added_to_dag : t -> Dag.Id.Set.t
+
+  val record_child_added_to_dag : t -> dag_node_id:Dag.Id.t -> unit
+end = struct
+  type phase =
+    | Restore_from_cache
+    | Compute
+
   type ('i, 'o) unpacked =
-    { without_state : ('i, 'o) Dep_node_without_state.t
-    ; running_state : Running_state.t
+    { dep_node : ('i, 'o) Dep_node_without_state.t
+    ; dag_node : Dag.node Lazy.t
+    ; (* CR-soon amokhov: Benchmark if it's worth switching to [Dag.Id.Table.t]. *)
+      mutable children_added_to_dag : Dag.Id.Set.t
```

Comment:
This table serves dual purpose:

1. guarantee that we never add the same edge twice to the cycle-detection DAG
2. mark the "forcing" stacks in the computation DAG that we already added

Consider that we have up to two stacks per cycle-detection DAG node, so it's not an idle question if something is per-dag-node or per-stack.

I think it's pretty clear that (1) needs to be per-cycle-detection-DAG-node, while (2) needs to be per-computation (so per-stack).

By making them both per-stack I'm suspecting you're opening the door to a bug where we can end up adding the same edge twice. (maybe we can have X -[validate]-> A -[validate]-> B followed by Y -[compute]-> A -[compute]-> B ?
Then the edge A->B will be added twice)

This is, of course, in addition to my pre-existing distaste to abusing the hashtable non-emptiness as a meaningful bool. :-)

I wonder, maybe it's simpler to get rid of Dag_node reuse than to be careful about fix this. I think if we do, then maybe we won't be needing `children_added_to_dag` at all?

After all, the forcing edges form a tree, not an arbitrary DAG (we have a unique incoming edge for each node).

The only reason we need it is because we can have two different stacks re-use the same node and therefore add the same edge.

Reply:
Indeed, I am now tempted to get rid of the DAG node reuse right in this PR because it would just simplify the reasoning a lot. I didn't want to conflate too many changes here initially, but perhaps it's not too bad. Happy to do that. What do you think?

Reply:
On the one hand, it's weird to merge with known issues. On the other, the damage from this bug will be limited even if we manage to hit it, somehow.
So I think it's OK if you prefer to merge this PR and then get rid of node re-use later.

Reply:
I'm afraid I'm wrong about our ability to completely get rid of `children_added_to_dag`. I think it's true that this table wouldn't be needed for forcing edges alone (because of the argument above), but we still need it for the non-forcing ones. We can have many identical non-forcing edges, after all.

Reply:
Good point about duplicate non-forcing edges. I still wonder if we can somehow make use of our knowledge that forcing edges form trees. 

Anyway, let's get rid of DAG node reuse in a follow-up PR.